### PR TITLE
Don't delete projects, mark as deleted

### LINKED ISSUE #1340: Don't delete projects, mark as deleted
Deleting a project causes missing links in other projects and goals and is dangerous

Warn user of unarchived goals in that project and archive them too

### PR DESCRIPTION
**Pull request by LamaEats**
Before mark the project as deleted all project's goals, linked criterias, affected goal from the other projects and affected projects get changes:
- goals of current projects will mark as archived
- linked criteria will mark as deleted
- for affected goals will recaclute criteria score
- for affected projects will recalculate average project's score by affected goals

After this operations the current project will mark as deleted

Also in the deleting dialog user will pay attention that project have active goals and these will move to an archive

If one of these steps will complete with error then all change will be restore by db transaction

## PR includes

- [x] Feature

## Related issues

Resolve #1340 

## QA Instructions, Screenshots, Recordings
  
image

If current project has children projects then user will see this modal dialog
image

### CODE REVIEW

File: prisma/schema.prisma [hunk 1]
```diff
@@ -153,6 +153,7 @@ model Project {
   watchers     Activity[] @relation("projectWatchers")
   stargizers   Activity[] @relation("projectStargizers")
   averageScore Int?
+  deleted      Boolean?   @default(false)
```
Comment (9teen90nine) [inline prisma/schema.prisma:156]: can we use `archived` to match goals? it wold be such a pain to keep both

File: src/hooks/useProjectResource.ts [hunk 2]
```diff
@@ -51,7 +51,7 @@ export const useProjectResource = (id: string) => {
     const deleteProject = useCallback(
         // eslint-disable-next-line @typescript-eslint/no-unused-vars
         (cb: Callback) => async () => {
-            const res = await deleteMutation.mutateAsync(id);
+            const res = await deleteMutation.mutateAsync({ id });
```
Comment (9teen90nine) [inline src/hooks/useProjectResource.ts:54]: error handling?

File: trpc/queries/goals.ts [hunk 3]
```diff
@@ -341,6 +341,9 @@ export const goalDeepQuery = {
                 },
             },
         },
+        where: {
+            archived: { not: true },
```
Comment (9teen90nine) [inline trpc/queries/goals.ts:345]: you can use `const notArchived = { archived: { not: true } }`

and then use query like `where: notArchived` which is both DRY and looks clean

File: trpc/router/goal.ts [hunk 4]
```diff
@@ -1235,4 +1236,18 @@ export const goal = router({
             throw new TRPCError({ code: 'INTERNAL_SERVER_ERROR', message: String(error.message), cause: error });
         }
     }),
+    getListByProjectId: protectedProcedure.input(batchGoalByProjectIdSchema).query(async ({ input }) => {
+        try {
+            return prisma.goal.findMany({
+                where: {
+                    projectId: input.projectId,
+                    AND: {
+                        OR: [{ archived: false }, { archived: null }],
```
Comment (9teen90nine) [inline trpc/router/goal.ts:1245]: not true?

File: trpc/router/project.ts [hunk 5]
```diff
@@ -238,6 +242,9 @@ export const project = router({
                                 },
                             },
                         },
+                        {
+                            OR: [{ deleted: null }, { deleted: false }],
```
Comment (9teen90nine) [inline trpc/router/project.ts:246]: the same here as with `notArchived` - you can use a variable to clean up these copies.

**Review by 9teen90nine — COMMENTED** [review]
Some minor thoughts I think are worth a fix but otherwise great work!

File: trpc/queries/goals.ts [hunk 6]
```diff
@@ -472,6 +481,11 @@ export const calcAchievedWeight = (
 ): number => {
     const { achivedWithWeight, comletedWithoutWeight, anyWithoutWeight, allWeight } = list.reduce(
         (acc, value) => {
+            // `where` filter by `deleted` field doesn't work in *Many queries
+            if (value.deleted) {
```
Comment (9teen90nine) [inline trpc/queries/goals.ts:485]: shouldn't it be now `archived`?

File: src/utils/db.ts [hunk 7]
```diff
@@ -340,3 +340,116 @@ export const updateGoalWithCalculatedWeight = async (goalId: string) => {
...
+                });
+
+                // mark criterias with these goals as deleted
+                const updateCriterias = ctx.goalAchieveCriteria.updateMany({
+                    where: {
+                        goalIdAsCriteria: {
+                            in: goalIds.map(({ id }) => id),
+                        },
+                    },
+                    data: { deleted: true },
```
Comment (9teen90nine) [inline src/utils/db.ts:392]: also here?

File: trpc/queries/goals.ts (hunk 6 above)
Comment (LamaEats) [reply trpc/queries/goals.ts:485]: this field for criteria table, i wont rename this field in this pr :)

File: trpc/router/goal.ts (hunk 4 above)
Comment (awinogradov) [reply trpc/router/goal.ts:1245]: Must be fixed ;)

**Review by 9teen90nine — APPROVED** [review]

**Review by 9teen90nine — APPROVED** [review]

**Pull request merged**